cpp/src/lists/contains.cu

```diff
@@ -0,0 +1,162 @@
+#include <cudf/lists/contains.hpp>
+#include <cudf/lists/list_device_view.cuh>
+#include <cudf/lists/lists_column_device_view.cuh>
+#include <cudf/lists/lists_column_view.hpp>
+#include <cudf/scalar/scalar.hpp>
+#include <cudf/scalar/scalar_device_view.cuh>
+#include <cudf/utilities/type_dispatcher.hpp>
+#include <rmm/exec_policy.hpp>
+#include <type_traits>
+
+namespace cudf {
+namespace lists {
+
+namespace {
+
+auto CUDA_HOST_DEVICE_CALLABLE counting_iter(size_type n)
+{
+  return thrust::make_counting_iterator(n);
+}
+
+std::pair<rmm::device_buffer, size_type> construct_null_mask(
+  cudf::detail::lists_column_device_view const& d_lists,
+  cudf::scalar const& skey,
+  bool input_column_has_nulls,
+  rmm::cuda_stream_view stream,
+  rmm::mr::device_memory_resource* mr)
+{
+  using namespace cudf;
+  using namespace cudf::detail;
+
+  auto const skey_is_valid    = skey.is_valid(stream);
+  auto const output_has_nulls = !skey_is_valid || input_column_has_nulls;
+
+  return output_has_nulls
+           ? (skey_is_valid
+                ? cudf::detail::valid_if(counting_iter(0),
+                                         counting_iter(d_lists.size()),
+                                         [d_lists] __device__(auto const& row_index) {
+                                           auto list = cudf::list_device_view(d_lists, row_index);
+                                           if (list.is_null()) { return false; }
+                                           return thrust::none_of(
+                                             thrust::seq,
+                                             counting_iter(0),
+                                             counting_iter(list.size()),
+                                             [&list] __device__(auto const& i) {
+                                               return list.is_null(i);
+                                             });
+                                         })
+                : std::make_pair(cudf::create_null_mask(d_lists.size(), mask_state::ALL_NULL, mr),
+                                 d_lists.size()))
+           : std::make_pair(rmm::device_buffer{0, stream, mr}, size_type{0});
+}
+
+struct lookup_functor {
+  template <typename T, typename... Args>
+  std::enable_if_t<!cudf::is_numeric<T>() && !std::is_same<T, cudf::string_view>::value, void>
+  operator()(Args&&...) const
+  {
+    CUDF_FAIL("lists::contains() is only supported on numeric types and strings.");
+  }
+
+  template <typename T>
+  std::enable_if_t<cudf::is_numeric<T>() || std::is_same<T, cudf::string_view>::value, void>
+  operator()(cudf::detail::lists_column_device_view const& d_lists,
+             cudf::scalar const& skey,
+             cudf::mutable_column_device_view output_bools,
+             rmm::cuda_stream_view stream,
+             rmm::mr::device_memory_resource* mr) const
+  {
+    if (!skey.is_valid()) {
+      return;  // Null mask is already set. No search required.
```

Comment:
The `skey` is a constant so why even type-dispatch this functor if `skey` is invalid?

Reply:
D'oh! Thanks for pointing that out, @davidwendt.